Stabilize TryFrom and TryInto with a convert::Infallible empty enum

SimonSapin opened the pull request:
This is the plan proposed in https://github.com/rust-lang/rust/issues/33417#issuecomment-423073898

rust-highfive commented:
r? @rkruppe

(rust_highfive has picked a reviewer for you, use r? to override)

SimonSapin commented:
@rfcbot fcp merge

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/176347749) ([raw log](https://api.travis-ci.com/v3/job/176347749/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:04d689f0:start=1549634997151285181,finish=1549635146244413755,duration=149093128574
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:58:49]     Checking compiler_builtins v0.1.5
[00:58:50]  Documenting alloc v0.0.0 (/checkout/src/liballoc)
[00:58:55]     Finished release [optimized] target(s) in 35.92s
[00:58:56]  Documenting core v0.0.0 (/checkout/src/libcore)
[00:59:12] warning: `[never]` cannot be resolved, ignoring it...
[00:59:12]     |
[00:59:12]     |
[00:59:12] 529 | /// This enum has the same role as [the `!` “never” type][never],
[00:59:12]     |
[00:59:12] note: lint level defined here
[00:59:12]    --> src/libcore/lib.rs:63:9
[00:59:12]     |
[00:59:12]     |
[00:59:12] 63  | #![warn(intra_doc_link_resolution_failure)]
[00:59:12]     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:59:12]     = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`
[00:59:12] 
[00:59:12] warning: `[never]` cannot be resolved, ignoring it...
[00:59:12]     |
[00:59:12]     |
[00:59:12] 529 | /// This enum has the same role as [the `!` “never” type][never],
[00:59:12]     |
[00:59:12]     = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`
[00:59:12] 
[00:59:18]     Finished release [optimized] target(s) in 23.19s
---
[00:59:24]     Checking rustc_tsan v0.0.0 (/checkout/src/librustc_tsan)
[00:59:24]     Checking rustc_msan v0.0.0 (/checkout/src/librustc_msan)
[00:59:24]     Checking rustc_asan v0.0.0 (/checkout/src/librustc_asan)
[00:59:24]  Documenting std v0.0.0 (/checkout/src/libstd)
[00:59:31] error: `[never]` cannot be resolved, ignoring it...
[00:59:31]     |
[00:59:31]     |
[00:59:31] 529 | /// This enum has the same role as [the `!` “never” type][never],
[00:59:31]     |
[00:59:31] note: lint level defined here
[00:59:31]    --> src/libstd/lib.rs:209:9
[00:59:31]     |
[00:59:31]     |
[00:59:31] 209 | #![deny(intra_doc_link_resolution_failure)]
[00:59:31]     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:59:31]     = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`
[00:59:31] 
[00:59:31] error: `[never]` cannot be resolved, ignoring it...
[00:59:31]     |
[00:59:31]     |
[00:59:31] 529 | /// This enum has the same role as [the `!` “never” type][never],
[00:59:31]     |
[00:59:31]     = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`
[00:59:31] 
[00:59:31] error: Could not document `std`.
[00:59:31] error: Could not document `std`.
[00:59:31] 
[00:59:31] Caused by:
[00:59:31]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --crate-name std src/libstd/lib.rs --color always --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/doc --cfg 'feature="backtrace"' --cfg 'feature="backtrace-sys"' --cfg 'feature="compiler_builtins"' --cfg 'feature="compiler_builtins_c"' --cfg 'feature="default"' --cfg 'feature="panic-unwind"' --cfg 'feature="panic_unwind"' --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps --extern alloc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/liballoc-9e6c0311b71511c6.rmeta --extern backtrace_sys=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libbacktrace_sys-f13b165ed9b4dd57.rmeta --extern compiler_builtins=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-d54fe968dea87029.rmeta --extern core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libcore-84e5b9599b1b7754.rmeta --extern libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/liblibc-a69fda92b07aedd5.rmeta --extern panic_abort=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libpanic_abort-349a3e5cce9f18ee.rmeta --extern panic_unwind=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libpanic_unwind-d7c48504ff1056b6.rmeta --extern rustc_demangle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librustc_demangle-20814bca47e9a554.rmeta --extern rustc_asan=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librustc_asan-8bc4c4bf1d4ec8f0.rmeta --extern rustc_lsan=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librustc_lsan-0814373332ff833c.rmeta --extern rustc_msan=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librustc_msan-655667bc8a1522d0.rmeta --extern rustc_tsan=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librustc_tsan-5756fbda854daecf.rmeta --extern unwind=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libunwind-5e31c590860d0940.rmeta` (exit code: 1)
[00:59:31] 
[00:59:31] 
[00:59:31] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-Z" "unstable-options" "-p" "std" "--" "--markdown-css" "rust.css" "--markdown-no-toc" "--index-page" "/checkout/src/doc/index.md"
[00:59:31] 
[00:59:31] 
[00:59:31] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap doc
[00:59:31] Build completed unsuccessfully in 0:06:12
[00:59:31] Build completed unsuccessfully in 0:06:12
[00:59:31] Makefile:18: recipe for target 'all' failed
[00:59:31] make: *** [all] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0e0d8a6d
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Fri Feb  8 15:12:07 UTC 2019
---
travis_time:end:0b378bfe:start=1549638728629181498,finish=1549638728633904520,duration=4723022
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0103429f
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0a0f6eff
travis_time:start:0a0f6eff
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0503f1e2
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

SimonSapin commented on src/libcore/convert.rs:
```diff
@@ -496,3 +500,95 @@ impl AsRef<str> for str {
+/// With `Infallible` being an enum, this code is valid.
+/// However when `Infallible` becomes an alias for the never type,
+/// the two `impl`s will start to overlap
+/// and therefore will be disallowed by the language’s trait coherence rules.
```
I think we want to add something here that says “so, uh, don’t do this?”. Wording suggestions welcome.

dwijnand replied:
Is using one over the other preferable?  Either way, maybe the answer to that should be in the docs.

alexcrichton commented on src/libcore/convert.rs:
```diff
@@ -496,3 +500,96 @@ impl AsRef<str> for str {
+    }
+}
+
+use fmt;
```
Could this be moved up to the top of the file?

alexcrichton commented on src/libcore/convert.rs:
```diff
@@ -496,3 +500,96 @@ impl AsRef<str> for str {
+}
+
+#[stable(feature = "convert_infallible", since = "1.34.0")]
+impl Eq for Infallible {}
```
With `PartialEq` and `Eq` perhaps the `Ord`-style traits could be included as well?

scottmcm commented:
:tada: Huzzah! 🎉 

Hopefully it sticks this time.

bors commented:
:umbrella: The latest upstream changes (presumably #58389) made this pull request unmergeable. Please resolve the merge conflicts.

hellow554 commented on src/libcore/convert.rs:
```diff
@@ -499,3 +499,96 @@ impl AsRef<str> for str {
+///
+/// # Future compatibility
+///
+/// This enum has the same role as [the `!` “never” type][never],
```
Maybe using regular quotes instead? "never"

hellow554 commented on src/libcore/convert.rs:
````diff
@@ -499,3 +499,96 @@ impl AsRef<str> for str {
+/// pub type Infallible = !;
+/// ```
+///
+/// … and eventually deprecate `Infallible`.
````
Maybe using three plain dots

hellow554 commented on src/libcore/convert.rs:
```diff
@@ -499,3 +499,96 @@ impl AsRef<str> for str {
+///
+///
+/// However there is one case where `!` syntax can be used
+/// before `!` is stabilized as a full-fleged type: in the position of a function’s return type.
```
`'`

Nemo157 replied:
```suggestion
/// before `!` is stabilized as a full-fledged type: in the position of a function’s return type.
```

"full-fledged" still looks wrong to me as a NZer, but Merriam-Webster seems to imply that "fully-fledged" is only a British thing.

hellow554 commented on src/libcore/convert.rs:
```diff
@@ -499,3 +499,96 @@ impl AsRef<str> for str {
+/// With `Infallible` being an enum, this code is valid.
+/// However when `Infallible` becomes an alias for the never type,
+/// the two `impl`s will start to overlap
+/// and therefore will be disallowed by the language’s trait coherence rules.
```
`'`

SimonSapin commented:
@rust-lang/docs, any opinion on the use of non-ASCII v.s. ASCII punctuation in docs? I use the former out of habit.

GuillaumeGomez commented:
We use it everywhere else in the doc so please use the ascii version of it. :)

frewsxcv commented:
To my knowledge, the docs team has no agreed upon preference for smart quotes or ASCII quotes, so either is fine!

jethrogb commented:
It seems to me stabilizing the never type is close (#57012). Is it really necessary to do this now or can we wait just a little bit longer?

SimonSapin commented:
A couple weeks ago I was very hopeful that the never type would indeed be close. However since then another concern was added in https://github.com/rust-lang/rust/issues/57012, and we didn’t find consensus on the related type inference fallback change.

`TryFrom` was overdue when we first tried to stabilize it a year ago. There was “[not much needed to unblock this one](https://github.com/rust-lang/rust/issues/33417#issuecomment-325797552)” 18 months ago. I personally feel we’ve done enough of waiting just a little bit longer.

SimonSapin commented:
I’ve added `impl From<!> for Infallible`. I initially also added the reverse but had to remove it because caused unexpected errors like:
    
```rust
error[E0277]: the trait bound `!: std::convert::From<()>` is not satisfied
   --> src/librustc_metadata/encoder.rs:105:9
    |
105 |         self.emit_usize(seq.len)?;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<()>` is not implemented for `!`
    |
    = help: the following implementations were found:
              <! as std::convert::From<std::convert::Infallible>>
    = note: the trait is implemented for `()`. Possibly this error has been caused by changes to Rust's type-inference algorithm (see: https://github.com/rust-lang/rust/issues/48950 for more info). Consider whether you meant to use the type `()` here instead.
    = note: required by `std::convert::From::from`
```
    
I don’t understand why this error happens. If I’m reading the code correctly the return types of `emit_usize` and of the method that contains line 105 are both `Result<(), !>`, so the expansion of the `?` operator should involve `!: From<!>`, not `From<()>`.

Is this a type inference bug?

Centril commented on src/libcore/num/mod.rs:
```diff
@@ -4523,24 +4517,34 @@ impl TryFromIntError {
 
-#[unstable(feature = "try_from", issue = "33417")]
+#[stable(feature = "try_from", since = "1.34.0")]
+impl From<Infallible> for TryFromIntError {
```
@eddyb @arielb1 Can you double check that this impl isn't problematic re. coherence and adding `From<!> for T` later when `type Infallible = !;` (e.g. re. https://github.com/rust-lang/rust/issues/57012#issuecomment-460740678)? -- I think it should be OK but I'd like to be more certain.

SimonSapin replied:
Since this impl is in the standard library, we get to remove it at the same time as whenever we add `From<!> for T`. It’s only compat with impls in other crates that we need to be worried about.

SimonSapin commented:
r? @alexcrichton, I’ve added b2cf9a02b2bce6ed984b0c967e2daf0af7b13629 after you last approved the code changes.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit cf267540ebabdeac1f2045819cd6bac561017e29 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton cf267540ebabdeac1f2045819cd6bac561017e29 -->

Centril commented:
Since this is a stabilization PR, we want to get it in before master=>beta promotion, so: @bors p=50

bors commented:
:hourglass: Testing commit cf267540ebabdeac1f2045819cd6bac561017e29 with merge 2e616be96c6c14265a4f55ca59f90f143e8b36a1...

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/102202512)

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/180351881) ([raw log](https://api.travis-ci.com/v3/job/180351881/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Attempting to download s3://rust-lang-ci-sccache2/docker/1f08d22f2a5e39c33797b6e95be989c2f1c727b72a5f7eeb8d59fa21b0c4483b32db64e100e5a6edf7e52a2b4841266c0b6e400d243c0f6e6db4570618ce4831
[00:00:14] Attempting with retry: curl -y 30 -Y 10 --connect-timeout 30 -f -L -C - -o /tmp/rustci_docker_cache https://s3-us-west-1.amazonaws.com/rust-lang-ci-sccache2/docker/1f08d22f2a5e39c33797b6e95be989c2f1c727b72a5f7eeb8d59fa21b0c4483b32db64e100e5a6edf7e52a2b4841266c0b6e400d243c0f6e6db4570618ce4831
[00:00:14]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[00:00:14]                                  Dload  Upload   Total   Spent    Left  Speed
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

SimonSapin commented:
@bors retry

bors commented:
:hourglass: Testing commit cf267540ebabdeac1f2045819cd6bac561017e29 with merge 00aae71f503b1ab592f48de47dd30912f3858748...

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/102224151), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/22638592)
Approved by: alexcrichton
Pushing 00aae71f503b1ab592f48de47dd30912f3858748 to master...

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/102224151), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/22638592)
Approved by: alexcrichton
Pushing 00aae71f503b1ab592f48de47dd30912f3858748 to master...

jethrogb commented:
We're missing `impl From<Infallible> for TryFromSliceError`, see #60318
